### PR TITLE
Try assigning fetch to globalThis if global assignment fails

### PR DESCRIPTION
**Pull request by sebmarkbage**
In case it's a more modern yet rigid environment.

### CODE REVIEW

**Review by sophiebits — APPROVED** [review]
Why not try globalThis first?

File: scripts/flow/environment.js [hunk 1]
```diff
@@ -18,6 +18,8 @@ declare var __REACT_DEVTOOLS_GLOBAL_HOOK__: any; /*?{
   inject: ?((stuff: Object) => void)
 };*/
 
+declare var globalThis: Object;
+
```
Comment (gnoff) [inline scripts/flow/environment.js:22]: Isn't this like typing it as any when it is quite unsafe to do so? why not just suppress the flow warning where you attempt to reference it if you don't want to existence check it?

**Review by gnoff — APPROVED** [review]
left a nit/question. But I am confused why this is necessary though I presume you wouldn't have thought to do this if not for a practical example where this fails. I would have guessed if the global was frozen or something like that then a trick with globalThis wouldn't be successul. I just don't have a mental model for where the second attempt would ever be successful if the first failed.

**Comment by jdalton** [comment]
Why not replace user references to fetch with the React one, during build/transform, without overwriting the one on the global object?

**Comment by sebmarkbage** [comment]
This is specifically for all kinds of code/libraries that’s already written for fetch and not specifically for React necessarily. Lots of code gets it from the global object (eg window.fetch()) and get to the global object in different ways.

**Comment by sebmarkbage** [comment]
Also React itself doesn’t have a compiler anyway so no way to make it universal that way anyway.

**Comment by jdalton** [comment]
This is a React package though (so React centric). At glance this looks like something that could continue to be a package and then shipped with a companion babel (and friends) transform in much the same way JSX is translated. Folks can always bolt on the global if they really want to but doing it by default seems less than great.

**Comment by sebmarkbage** [comment]
It's really worth talking about when it actually kicks in, what it actually does even if it does kick, that it doesn't add any new methods or anything to the prototype and that it gracefully backs off if you're in a frozen environment.

If it was a separate package, everyone mistaking installing it in their custom runtimes would miss out. If it was highly recommended everywhere and everyone used it anyway, what is it protecting against?

Other libraries already do this less gracefully so whatever changes the platform would do is already busted. We're not the first by far.

Also, this is not done yet and there's a larger RFC to describe the whole story before people jump to conclusions.

**Comment by jdalton** [comment]
> It's really worth talking about when it actually kicks in, what it actually does even if it does kick in, that it doesn't add any new methods, options or anything to the prototype and that it gracefully backs off if you're in a frozen environment.

In frozen environments folks trying to rely on it couldn't where as if user code was transpiled with `fetch()` references replaced with the React flavor then they could still use this cacheable fetch.

> If it was highly recommended everywhere and everyone used it anyway, what is it protecting against?

Highly recommended in the React ecosystem is different than the entire platform.

> Other libraries already do this less gracefully so whatever changes the platform would do is already busted. We're not the first by far.

I mean yes, libraries have done this but it's not been great for the language/ecosystem. 

> Also, this is not done yet and there's a larger RFC to describe the whole story before people jump to conclusions.

Okay.

**Comment by sebmarkbage** [comment]
> Highly recommended in the React ecosystem is different than the entire platform.

Sorry I meant if we recommended the explicit patching in the entire React ecosystem vs implicit. If it was always separate import or compiled by scope it would be a different story, but that's not enough.

> libraries have done this but it's not been great for the language/ecosystem.

I mean specifically for `fetch()`. Hydration based libraries patch it to restore fetches that happened on the server.

**Comment by jdalton** [comment]
@sebmarkbage Do you have a reference to the larger RFC?

**Comment by iRunner** [comment]
https://github.com/acdlite/rfcs/blob/first-class-promises/text/0000-first-class-support-for-promises.md

Is this it?

**Comment by gaearon** [comment]
No, the relevant RFC isn’t finished or published yet.

**Pull request merged**